problemtools/verifyproblem.py

```diff
@@ -361,29 +381,45 @@ def check(self, args):
         return self._check_res
 
 
-    def compute_result(self, sub_results, probtype, on_reject, shadow_result=False):
-        grade = None
-        if probtype == 'scoring':
-            grade = lambda x: self._problem.graders.grade(x, self, shadow_result)
-        return SubmissionResult.aggregate_results(sub_results, on_reject, grade=grade)
-
-
     def run_submission(self, sub, args, timelim_low, timelim_high):
         self.info('Running on %s' % self)
         subres1 = []
         subres2 = []
-        probtype = self._problem.config.get('type')
-        on_reject = self._problem.config.get('grading')['on_reject']
+        on_reject = self.config['on_reject']
         for subdata in self._items:
             if not subdata.matches_filter(args.data_filter):
                 continue
             (r1, r2) = subdata.run_submission(sub, args, timelim_low, timelim_high)
             subres1.append(r1)
             subres2.append(r2)
-            if on_reject == 'first_error' and r2.verdict != 'AC':
+            if on_reject == 'break' and r2.verdict != 'AC':
                 break
-        return (self.compute_result(subres1, probtype, on_reject),
-                self.compute_result(subres2, probtype, on_reject, shadow_result=True))
+        return (self.aggregate_results(subres1),
+                self.aggregate_results(subres2, shadow_result=True))
+
+    def aggregate_results(self, sub_results, shadow_result=False):
+        res = SubmissionResult(None)
+
+        for r in sub_results:
+            if r.runtime > res.runtime:
+                res.runtime = r.runtime
+                res.runtime_testcase = r.runtime_testcase
+            if r.ac_runtime > res.ac_runtime:
+                res.ac_runtime = r.ac_runtime
+                res.ac_runtime_testcase = r.ac_runtime_testcase
+
+        judge_error = next((r for r in sub_results if r.verdict == 'JE'), None)
+        if judge_error:
+            res.verdict = judge_error.verdict
+            res.reason = judge_error.reason
+            res.testcase = judge_error.testcase
+        else:
+            (res.verdict, score) = self._problem.graders.grade(sub_results, self, shadow_result)
+            if self._problem.config.get('type') == 'scoring':
+                res.score = score
+            res.testcase = sub_results[-1].testcase
```

Comment:
Essentially it ensures res.testcase is always the last testcase ran for a certain subresult, which imho seems like a reasonable definition for what the a testcase property for a result would be.